src/client_side_reply.cc

```diff
@@ -1562,17 +1559,10 @@ clientReplyContext::buildReplyHeader()
         Auth::UserRequest::AddReplyAuthHeader(reply, request->auth_user_request, request, http->flags.accel, 0);
 #endif
 
-    /* Append X-Cache */
-    httpHeaderPutStrf(hdr, Http::HdrType::X_CACHE, "%s from %s",
-                      is_hit ? "HIT" : "MISS", getMyHostname());
-
-#if USE_CACHE_DIGESTS
-    /* Append X-Cache-Lookup: -- temporary hack, to be removed @?@ @?@ */
-    httpHeaderPutStrf(hdr, Http::HdrType::X_CACHE_LOOKUP, "%s from %s:%d",
-                      lookup_type ? lookup_type : "NONE",
-                      getMyHostname(), getMyPort());
-
-#endif
+    httpHeaderPutStrf(hdr, Http::HdrType::CACHE_STATUS, "%s;%s",
+                     uniqueHostname(),
+                     (is_hit ? "hit" : "fwd=miss")
```

Comment:
> Alex: Should we drop the X-Cache-Lookup as well?
> Amos: I skipped it because it reports HIT and MISS wrongly.

Hm... I cannot find a reasonable interpretation of that answer, but I am OK with X-Cache-Lookup removal in this PR. Please do not forget to add the removed _information_ to Cache-Status though!

FWIW, X-Cache-Lookup is not about a cache hit or miss. X-Cache-Lookup is about initial cache index query -- it is a triage aid that cannot imply a (future) cache hit because the cache checks some but not all the response matching criteria, and because a lot of things can still go wrong even after a full match; a negative (i.e. "not found") result implies a miss AFAICT.

An Optional boolean field (clientReplyContext::positiveCacheLookup?) may work as a lookup_type replacement.

Reply:
> FWIW, X-Cache-Lookup is not about a cache hit or miss.

That is absurd. The header quite literally says "HIT from proxy.example.org:3128" or "MISS from proxy.example.org:3128" with the words being based on a cache lookup HIT/MISS result.


> X-Cache-Lookup is about initial cache index query -- it is a triage aid that cannot imply a (future) cache hit because the cache checks some but not all the response matching criteria, and because a lot of things can still go wrong even after a full match; a negative (i.e. "not found") result implies a miss AFAICT.
> 
> An Optional boolean field (clientReplyContext::positiveCacheLookup?) may work as a lookup_type replacement.

Those additional criteria inclientReplyContext::identifyFoundObject() are possibly a good basis for selecting some of the other fwd= values this header could send. But transferring that information around involves changing the APIs and testing the result accuracy - which is out of the scope selected for this PR.

Reply:
> > FWIW, X-Cache-Lookup is not about a cache hit or miss.
 
> That is absurd.

It is a fact. Whether you consider it absurd may explain your misunderstanding but does not change the meaning of this header.

> The header quite literally says "HIT from proxy.example.org:3128" or "MISS from proxy.example.org:3128" with the words being based on a cache lookup HIT/MISS result.

While I agree that the choice of the overloaded `HIT` and `MISS` terms was a bit unfortunate for this context, they correctly convey the _lookup_ outcome. Please note that the header field name tells us that its value is about the _lookup_, not whether the transaction is classified as a cache hit or cache miss.

 
> > X-Cache-Lookup is about initial cache index query -- it is a triage aid that cannot imply a (future) cache hit because the cache checks some but not all the response matching criteria, and because a lot of things can still go wrong even after a full match; a negative (i.e. "not found") result implies a miss AFAICT.
> > An Optional boolean field (clientReplyContext::positiveCacheLookup?) may work as a lookup_type replacement.
 
> Those additional criteria inclientReplyContext::identifyFoundObject() are possibly a good basis for selecting some of the other fwd= values this header could send. But transferring that information around involves changing the APIs and testing the result accuracy - which is out of the scope selected for this PR.

This change request does not require any API changes. It requires a tiny adjustment of the existing code to preserve useful information currently relayed by the being-removed X-Cache-Lookup. Please let me know if I should implement this myself. I think I found a way to do that with even fewer changes than using Optional would require.

Reply:
So do you know why it often says "HIT" on responses from proxies that have no cache to look up? and "MISS" on responses obviously pulled from cache? Under your definition of its meaning both should be impossible states.
The only case where I see it conveying truth and being different from X-Cache is when a MISS is happening due to Vary.

Yes, please implement this solution you have. It would be faster than debating this further.

Reply:
> do you know why it often says "HIT" on responses from proxies that have no cache to look up?

I can offer one (speculative!) suspect: An object with the same key was in the `store_table` at the time of the request. Unfortunately, Squid uses store_table for all transactions, regardless of whether caching is enabled. I hope to fix that old design bug in the foreseeable future.

> and "MISS" on responses obviously pulled from cache?

I know of one case where the initial lookup_type `HIT` value gets overwritten but it does not match the "pulled from cache" precondition in your question: When Squid finds a matching cached response with a Vary response header, identifyFoundObject() sets lookup_type to `HIT`. Squid then tries to find another response object using a Vary-augmented Store key. That attempt results in the second identifyFoundObject() call (which lookup_type authors could not expect because Squid did not know about Vary at the time). That second call may change the lookup_type `HIT` value to `MISS`.


> Under your definition of its meaning both should be impossible states.

... or Squid bugs or our insufficient understanding of the corresponding (very convoluted!) code. Investigating a couple of such cases may be quite useful. FWIW, I have not noticed any such cases, but I was not looking for them.

BTW, the X-Cache:MISS value is often lying in delay pools environments due to the known [bug](http://lists.squid-cache.org/pipermail/squid-users/2019-March/020388.html). It would be very nice to fix that bug _before_ merging Cache-Status, so that we know that all Cache-Status responses do not suffer from that annoying and dangerous problem! Please see my squid-bugs email for more details.


> Yes, please implement this solution you have.

Done in dda920d. Please review and adjust as needed. Here is sample output:


```
HTTP/1.1 503 Service Unavailable
X-Squid-Error: ERR_CONNECT_FAIL 111
Vary: Accept-Language
Cache-Status: protofroot;fwd=miss;detail=match
```